examples/4_totp_storage.py

```diff
@@ -0,0 +1,218 @@
+    for login processes that require two-step verification. This bot uses the
+    pyotp library, so it is also compatible with other MFA apps.
+
+    A Keybase TOTP bot could be useful for logins where you're already using a strong
+    password, but are required to also use TOTP. Or it could be perfect for adding
+    replay attack resistance to the physical keypad lock protecting your team's ice
+    cream freezer.... Note that depending on the login process and how you use this bot
+    with your Keybase team setup, this bot may not be appropriate for threat models
+    that require a physical "second factor" device.
+
+    For more information on TOTP, see https://pyotp.readthedocs.io/en/latest/,
+    https://tools.ietf.org/html/rfc6238.
+    """
+
+    MSG_PREFIX = "!totp"
+    NAMESPACE = "totp"
+
+    def to_json(self, secret):
+        return {"secret": secret}
+
+    async def __provision(self, bot, team, issuer, force: bool = False):
+        secret = pyotp.random_base32()
+        val = json.dumps(self.to_json(secret))
+        if not force:
+            try:
+                # throws exception if 1 is not the latest revision + 1
+                await bot.kvstore.put(team, self.NAMESPACE, issuer, val, revision=1)
+            except Exception as e:
+                res = await bot.kvstore.get(team, self.NAMESPACE, issuer)
+                if bot.kvstore.is_deleted(res):
+                    # then TOTP for this issuer was previously provisioned but
+                    # was deleted. insert as latest revision + 1
+                    await bot.kvstore.put(team, self.NAMESPACE, issuer)
+                else:
+                    # then TOTP for this issuer was previously provisioned and
+                    # might still be in use, so raise the exception
+                    raise e
+        else:
+            # reprovisioning; "force" provisioning by inserting as latest revision + 1
+            await bot.kvstore.put(team, self.NAMESPACE, issuer, val)
+        return pyotp.TOTP(secret).provisioning_uri(team, issuer_name=issuer)
+
+    async def __remove(self, bot, team, issuer):
+        # throws exception if nothing to delete
+        await bot.kvstore.delete(team, self.NAMESPACE, issuer)
+
+    async def __list(self, bot, team):
+        # returns all namespaces in this team; assumes that all namespaces are
+        # used for storing TOTP credentials
+        res = await bot.kvstore.list_entrykeys(team, self.NAMESPACE)
+        if res.entry_keys:
+            return [e.entry_key for e in res.entry_keys]
+        else:
+            return []
+
+    async def __uri(self, bot, team, issuer):
+        totp = await self.__totp(bot, team, issuer)
+        return totp.provisioning_uri(team, issuer_name=issuer) if totp else None
+
+    async def __now(self, bot, team, issuer):
+        totp = await self.__totp(bot, team, issuer)
+        return totp.now() if totp else None
+
+    async def __totp(self, bot, team, issuer):
```

Comment:
i'm curious about your decision to use dunder names for all of these methods. i was always told not to use dunders in cases like these because they convey that you're writing something intended to be applied at a lower level (e.g.  implementing __next__ for a custom iterable). i would have expected all of these to be single underscored since they seem much more like private methods.

Reply:
i was going for private method syntax (https://linux.die.net/diveintopython/html/object_oriented_framework/private_functions.html). but i'm also on the fence about the necessity and readability of using the "__" prefix.